In [1]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [2]:
df = pd.read_parquet("../data/processed/fbref_tm_eda.parquet")

In [3]:
df.head()

,Rk,Player,Nation,Pos,Squad,Comp,Age,Born,Playing Time_MP,Playing Time_Starts,Playing Time_Min,Playing Time_90s,Performance_Gls,Performance_Ast,Performance_G+A,Performance_G-PK,Performance_PK,Performance_PKatt,Performance_CrdY,Performance_CrdR,Per 90 Minutes_Gls,Per 90 Minutes_Ast,Per 90 Minutes_G+A,Per 90 Minutes_G-PK,Per 90 Minutes_G+A-PK,90s,Standard_Gls,Standard_Sh,Standard_SoT,Standard_SoT%,Standard_Sh/90,Standard_SoT/90,Standard_G/Sh,Standard_G/SoT,Standard_PK,Standard_PKatt,Performance_2CrdY,Performance_Fls,Performance_Fld,Performance_Off,Performance_Crs,Performance_Int,Performance_TklW,Performance_OG,Playing Time_Mn/MP,Playing Time_Min%,Starts_Starts,Starts_Mn/Start,Starts_Compl,Subs_Subs,Subs_Mn/Sub,Subs_unSub,Team Success_PPM,Team Success_onG,Team Success_onGA,Team Success_+/-,Team Success_+/-90,Team Success_On-Off,Season,nom,saison_id,player_id,anio_nac,pid_fb,valor_eur,posicion_tm,altura_m,nacimiento,club,contrato_vence,club_coincide,flag_sin_tiros,flag_sin_titularidades,flag_sin_suplencias,flag_cruce_verificado,flag_contrato_conocido,dias_contrato_restante,log_valor_eur,n_metodos_atipico,log1p_Performance_Gls,log1p_Performance_Ast,log1p_Standard_Sh,log1p_Standard_SoT,log1p_Performance_Crs,log1p_Performance_Fls,Age_rs,altura_m_rs,Playing Time_Min_rs,Performance_Gls_rs,Performance_Ast_rs,Standard_Sh_rs,Performance_CrdY_rs,Team Success_PPM_rs
0,1,Max Aarons,ENG,DF,Norwich City,Premier League,21.0,2000.0,34,32,2881,32.0,0,2,2,0,0,0,8,0,0.00,0.06,0.06,0.00,0.06,32.0,0,13,2,15.4,0.41,0.06,0.00,0.00,0,0,0,26,52,1,50,26,45,1,85,84.2,32,89.0,29,2,22.0,2,0.53,18,69,-51,-1.59,0.08,2021-2022,max aarons,2021,471690,2000.0,471690,22000000.0,Right-Back,1.77,2000-01-04,Norwich City,NaT,True,False,False,False,True,False,NaN,7.342423,0,0.000000,1.098612,2.639057,1.098612,3.931826,3.295837,-0.666667,-0.6,1.055258,0.0,1.0,0.20,1.50,-0.949367
1,2,Yunis Abdelhamid,MAR,DF,Reims,Ligue 1,33.0,1987.0,34,34,2983,33.1,2,0,2,2,0,0,5,1,0.06,0.00,0.06,0.06,0.06,33.1,2,18,6,33.3,0.54,0.18,0.11,0.33,0,0,0,38,25,0,4,67,28,0,88,87.2,34,88.0,33,0,0.0,0,1.12,38,41,-3,-0.09,-0.50,2021-2022,yunis abdelhamid,2021,199704,1987.0,199704,1200000.0,Centre-Back,1.90,1987-09-28,Stade Reims,2022-06-30,True,False,False,True,True,True,333.0,6.079181,0,1.098612,0.000000,2.944439,1.945910,1.609438,3.663562,1.333333,0.7,1.115865,1.0,0.0,0.45,0.75,-0.202532
2,3,Salis Abdul Samed,GHA,MF,Clermont Foot,Ligue 1,21.0,2000.0,31,29,2462,27.4,1,0,1,1,0,0,12,3,0.04,0.00,0.04,0.04,0.04,27.4,1,18,5,27.8,0.66,0.18,0.06,0.20,0,0,2,43,37,0,14,40,26,0,79,72.0,29,82.0,19,2,38.0,1,1.10,26,43,-17,-0.62,0.69,2021-2022,salis abdul samed,2021,688707,2000.0,688707,3000000.0,Defensive Midfield,1.79,2000-03-26,Clermont Foot 63,2025-06-30,True,False,False,False,True,True,1429.0,6.477121,0,0.693147,0.000000,2.944439,1.791759,2.708050,3.784190,-0.666667,-0.4,0.806298,0.5,0.0,0.45,2.50,-0.227848
3,4,Laurent Abergel,FRA,MF,Lorient,Ligue 1,28.0,1993.0,34,34,2956,32.8,0,2,2,0,0,0,9,0,0.00,0.06,0.06,0.00,0.06,32.8,0,30,7,23.3,0.91,0.21,0.00,0.00,0,0,0,41,69,1,37,54,63,0,87,86.4,34,87.0,26,0,0.0,0,0.82,27,59,-32,-0.97,-1.75,2021-2022,laurent abergel,2021,238626,1993.0,238626,3000000.0,Defensive Midfield,1.70,1993-02-01,FC Lorient,2025-06-30,True,False,False,True,True,True,1429.0,6.477121,0,0.000000,1.098612,3.433987,2.079442,3.637586,3.737670,0.500000,-1.3,1.099822,0.0,1.0,1.05,1.75,-0.582278
4,5,Charles Abi,FRA,FW,Saint-Étienne,Ligue 1,21.0,2000.0,1,1,45,0.5,0,0,0,0,0,0,0,0,0.00,0.00,0.00,0.00,0.00,0.5,0,0,0,0.0,0.00,0.00,0.00,0.00,0,0,0,1,0,0,1,0,0,0,45,1.3,1,45.0,0,0,0.0,0,1.00,0,0,0,0.00,0.93,2021-2022,charles abi,2021,418658,2000.0,418658,1500000.0,Centre-Forward,1.89,2000-04-12,AS Saint-Étienne,NaT,True,True,False,True,True,False,NaN,6.176091,0,0.000000,0.000000,0.000000,0.000000,0.693147,0.693147,-0.666667,0.6,-0.629828,0.0,0.0,-0.45,-0.50,-0.354430


In [4]:
CHAMPIONS = {
    2021: "Real Madrid",
    2022: "Manchester City",
    2023: "Real Madrid",
    2024: "PSG",
    2025: "PSG",
}

LIGAS = {
    ("Premier League", 2021): "Manchester City",
    ("Premier League", 2022): "Manchester City",
    ("Premier League", 2023): "Manchester City",
    ("Premier League", 2024): "Liverpool",
    ("Premier League", 2025): "Arsenal",
    ("La Liga", 2021): "Real Madrid",
    ("La Liga", 2022): "Barcelona",
    ("La Liga", 2023): "Real Madrid",
    ("La Liga", 2024): "Barcelona",
    ("La Liga", 2025): "Barcelona",
    ("Bundesliga", 2021): "Bayern Munich",
    ("Bundesliga", 2022): "Bayern Munich",
    ("Bundesliga", 2023): "Leverkusen",
    ("Bundesliga", 2024): "Bayern Munich",
    ("Bundesliga", 2025): "Bayern Munich",
    ("Serie A", 2021): "Milan",
    ("Serie A", 2022): "Napoli",
    ("Serie A", 2023): "Inter",
    ("Serie A", 2024): "Napoli",
    ("Serie A", 2025): "Inter",
    ("Ligue 1", 2021): "PSG",
    ("Ligue 1", 2022): "PSG",
    ("Ligue 1", 2023): "PSG",
    ("Ligue 1", 2024): "PSG",
    ("Ligue 1", 2025): "PSG",
}

df['Ganador_Champions'] = [int(CHAMPIONS.get(t) == s) for t, s in zip(df['saison_id'], df['Squad'])]
df['Ganador_Liga'] = [int(LIGAS.get((c, t)) == s)for c, t, s in zip(df['Comp'], df['saison_id'], df['Squad'])]

# Ingeniería de variables para predecir `valor_eur`

Partimos del dataset que dejó el EDA (`02-eda.ipynb`): sin duplicados raros, nulos ya imputados con sus flags, atípicos conservados a propósito, y algunas columnas ya normalizadas (`_rs`, `log1p_`).

Acá dejamos `df` listo como matriz de variables para el modelo: corregimos un bug de cruce que encontramos después, sacamos identificadores/columnas de auditoría y una fuga de información, eliminamos duplicados y derivadas, decidimos una sola representación por variable, y armamos las dummies de posición/liga/equipo. Cada paso de código va con su celda explicando el porqué.

## 1. Qué trae `df` heredado del EDA

Antes de tocar nada, miramos tipos de dato, nulos que quedan, y qué columnas están duplicadas o son combinación lineal de otras -- redundancia exacta, no solo correlación alta.

In [5]:
print(f"Shape inicial: {df.shape}")
print()
print("Nulos restantes:")
print(df.isna().sum().pipe(lambda s: s[s > 0]))
print()
print("Cardinalidad de columnas categóricas / booleanas:")
for c in df.columns:
    if df[c].dtype.name in ("category", "object", "bool"):
        print(f"  {c:26s} {df[c].nunique():>4d} categorías")

Shape inicial: (13835, 95)

Nulos restantes:
anio_nac                   686
pid_fb                     686
nacimiento                   2
contrato_vence            5939
dias_contrato_restante    5939
dtype: int64

Cardinalidad de columnas categóricas / booleanas:
  Nation                      134 categorías
  Pos                          10 categorías
  Squad                       130 categorías
  Comp                          5 categorías
  Season                        5 categorías
  posicion_tm                  13 categorías
  club                        130 categorías
  club_coincide                 2 categorías
  flag_sin_tiros                2 categorías
  flag_sin_titularidades        2 categorías
  flag_sin_suplencias           2 categorías
  flag_cruce_verificado         2 categorías
  flag_contrato_conocido        2 categorías


## 2. Arreglando un bug de cruce, y sacando columnas que no sirven de feature

### 2.1 Cuando dos jugadores distintos quedan pegados

Esto lo encontramos mirando los residuos de un Random Forest en `04-modelado.ipynb`: las predicciones para "Rodri" no tenían sentido entre filas que se supone tenían el mismo `valor_eur`. La causa está en el cruce de `01-unificacion.ipynb` -- se hace por nombre, y a veces dos jugadores distintos con el mismo nombre normalizado terminan compartiendo el mismo `player_id` de Transfermarkt (y su valor de mercado).

Afecta a 735 grupos jugador-temporada (~10.6% del dataset). Para resolverlo usamos `club_coincide` (¿el `Squad` de FBref coincide con el club de Transfermarkt?): si una sola fila del grupo coincide, esa es la buena y tiramos el resto; si ninguna coincide, no hay forma confiable de saber cuál es la correcta y se descarta el grupo entero.

**Vuelta después de entregar el proyecto:** un jugador cedido a otro club también genera dos `Squad` en la misma temporada, pero ahí NO es una colisión -- es la misma persona. `club_coincide` no distinguía esto, y como el club de Transfermarkt siempre es el dueño de la ficha (nunca el que recibe la cesión), terminaba tirando la fila de la cesión. Nos dimos cuenta con Endrick (cedido todo 2025-2026 en el Olympique Lyon, 16 partidos y 5 goles) -- se quedaba solo con los 12 minutos que jugó en el Real Madrid después de volver. Arreglamos esto con `scripts/detectar_prestamos.py`, que lee directamente las insignias de cesión de las plantillas de Transfermarkt y, cuando hay evidencia de que fue cedido, prioriza esa fila por sobre `club_coincide`.

### 2.2 Identificadores y columnas de auditoría

| Columna | Por qué se va |
|---|---|
| `Rk` | Índice del scraping, no dice nada del jugador |
| `nom`, `pid_fb`, `anio_nac` | Solo sirvieron para el cruce de fuentes, ya cumplieron su función |
| `Born`, `nacimiento` | Redundantes con `Age` |
| `club` | Es una foto del club actual, no un histórico -- usamos `Squad` |
| `Season` | Redundante con `saison_id` |
| `flag_cruce_verificado`, `club_coincide` | Auditoría del cruce, no un atributo del jugador |
| `n_metodos_atipico` | **Fuga de información** -- se calculó usando `valor_eur` como input en el EDA |

`player_id` y `Player` se quedan, pero solo como identificadores (para trazabilidad y para separar train/test por jugador), nunca como feature del modelo.

In [6]:
import re
import unicodedata


def normalizar_club(s):
    s = unicodedata.normalize("NFKD", str(s))
    s = "".join(c for c in s if not unicodedata.combining(c))
    s = s.replace("đ", "d").replace("Đ", "D").replace("ł", "l").replace("ø", "o")
    s = re.sub(r"[^a-z ]", "", s.lower())
    return re.sub(r"\s+", " ", s).strip()


def coincide(a, b):
    a, b = normalizar_club(a), normalizar_club(b)
    return (a in b) or (b in a)


# Evidencia de cesión (insignias "On loan from X" / "Returned after loan spell with X" de
# las plantillas de Transfermarkt, ver scripts/detectar_prestamos.py). club_coincide por sí
# solo siempre prefiere al club DUEÑO de la ficha -- para un jugador cedido, eso descarta
# sistemáticamente la fila de la cesión (a veces su temporada entera), aunque ambas filas
# sean la MISMA persona, no una colisión de nombre. Caso encontrado por el usuario:
# Endrick, cedido en Olympique Lyon toda 2025-2026 (16 partidos, 5 goles) -- club_coincide
# se quedaba con la fila de Real Madrid, sus 12 minutos post-regreso de la cesión.
prestamos = pd.read_csv("../data/interim/prestamos.csv", dtype={"player_id": str})
club_prestamo_de = dict(zip(zip(prestamos["player_id"], prestamos["saison_id"]), prestamos["club_prestamo"]))
club_prestamo = [club_prestamo_de.get((pid, sid)) for pid, sid in zip(df["player_id"], df["saison_id"])]
df["_es_fila_prestamo"] = [
    cp is not None and coincide(sq, cp) for sq, cp in zip(df["Squad"], club_prestamo)
]

key_cols = ["player_id", "saison_id"]
n_squads = df.groupby(key_cols)["Squad"].transform("nunique")
ambiguos = n_squads > 1
tiene_prestamo = df.groupby(key_cols)["_es_fila_prestamo"].transform("any")
n_true_grupo = df.groupby(key_cols)["club_coincide"].transform("sum")

n_grupos_ambiguos = df.loc[ambiguos, key_cols].drop_duplicates().shape[0]
n_grupos_prestamo = df.loc[ambiguos & tiene_prestamo, key_cols].drop_duplicates().shape[0]
n_grupos_resueltos = df.loc[ambiguos & ~tiene_prestamo & (n_true_grupo == 1), key_cols].drop_duplicates().shape[0]
n_grupos_sin_confianza = df.loc[ambiguos & ~tiene_prestamo & (n_true_grupo == 0), key_cols].drop_duplicates().shape[0]

print(f"Grupos (player_id, saison_id) con Squad ambiguo: {n_grupos_ambiguos}")
print(f"  - resueltos via evidencia de cesión (queda la fila del club donde jugó): {n_grupos_prestamo}")
print(f"  - resueltos via club_coincide (queda 1 fila correcta): {n_grupos_resueltos}")
print(f"  - sin ninguna fila confiable, se descarta el grupo completo: {n_grupos_sin_confianza}")

descartar_club_coincide = (
    (ambiguos & (n_true_grupo == 1) & (~df["club_coincide"]))
    | (ambiguos & (n_true_grupo == 0))
)
descartar_prestamo = ambiguos & tiene_prestamo & (~df["_es_fila_prestamo"])
descartar = np.where(tiene_prestamo, descartar_prestamo, descartar_club_coincide)

print(f"\nFilas descartadas: {descartar.sum()} de {df.shape[0]}")
df = df[~descartar].reset_index(drop=True)
df = df.drop(columns=["_es_fila_prestamo"])
print(f"Shape actual: {df.shape}")

Grupos (player_id, saison_id) con Squad ambiguo: 735
  - resueltos via evidencia de cesión (queda la fila del club donde jugó): 58
  - resueltos via club_coincide (queda 1 fila correcta): 603
  - sin ninguna fila confiable, se descarta el grupo completo: 74

Filas descartadas: 812 de 13835
Shape actual: (13023, 95)


In [7]:
cols_identificadores_auditoria = [
    "Rk", "nom", "pid_fb", "anio_nac", "Born", "nacimiento", "club", "Season",
    "flag_cruce_verificado", "club_coincide",
]
cols_leakage = ["n_metodos_atipico"]

df = df.drop(columns=cols_identificadores_auditoria + cols_leakage)
print(f"Eliminadas {len(cols_identificadores_auditoria)} columnas de identificadores/auditoría "
      f"y {len(cols_leakage)} columna con fuga de información.")
print(f"Shape actual: {df.shape}")

Eliminadas 10 columnas de identificadores/auditoría y 1 columna con fuga de información.
Shape actual: (13023, 84)


## 3. Duplicados exactos y variables derivadas

FBref repite algunas métricas con nombres distintos en sus tablas. Confirmamos que son idénticas (o suma/resta exacta de otras que ya tenemos) antes de tirarlas -- no aportan nada nuevo, solo multicolinealidad:

| Columna que se va | Es igual a... |
|---|---|
| `Standard_Gls` | `Performance_Gls` |
| `Standard_PK` / `Standard_PKatt` | `Performance_PK` / `Performance_PKatt` |
| `90s` | `Playing Time_90s` |
| `Starts_Starts` | `Playing Time_Starts` |
| `Performance_G+A` | `Gls` + `Ast` |
| `Performance_G-PK` | `Gls` − `PK` |
| `Per 90 Minutes_G+A`, `G-PK`, `G+A-PK` | versiones por-90 de lo mismo |
| `Team Success_+/-` y `+/-90` | `onG` − `onGA`, normalizado |

Nos quedamos con una sola versión de cada una.

In [8]:
# Verificación de igualdad exacta antes de eliminar
dup_pairs = [
    ("Standard_Gls", "Performance_Gls"),
    ("Standard_PK", "Performance_PK"),
    ("Standard_PKatt", "Performance_PKatt"),
    ("90s", "Playing Time_90s"),
    ("Starts_Starts", "Playing Time_Starts"),
]
for a, b in dup_pairs:
    print(f"{a:18s} == {b:22s} -> {(df[a] == df[b]).mean():.3f} de las filas iguales")

cols_duplicadas = [a for a, _ in dup_pairs]
cols_derivadas = [
    "Performance_G+A", "Performance_G-PK",
    "Per 90 Minutes_G+A", "Per 90 Minutes_G-PK", "Per 90 Minutes_G+A-PK",
    "Team Success_+/-", "Team Success_+/-90",
]

df = df.drop(columns=cols_duplicadas + cols_derivadas)
print(f"\nEliminadas {len(cols_duplicadas)} columnas duplicadas y {len(cols_derivadas)} derivadas.")
print(f"Shape actual: {df.shape}")

Standard_Gls       == Performance_Gls        -> 1.000 de las filas iguales
Standard_PK        == Performance_PK         -> 1.000 de las filas iguales
Standard_PKatt     == Performance_PKatt      -> 1.000 de las filas iguales
90s                == Playing Time_90s       -> 1.000 de las filas iguales
Starts_Starts      == Playing Time_Starts    -> 1.000 de las filas iguales

Eliminadas 5 columnas duplicadas y 7 derivadas.
Shape actual: (13023, 72)


## 4. Cinco formas de medir lo mismo: cuánto jugó

`Playing Time_MP`, `Starts`, `90s` y `Starts_Compl` correlacionan >0.88 con `Playing Time_Min` -- son casi la misma cosa. Nos quedamos solo con `Playing Time_Min` (su versión escalada, `_rs`) y tiramos las demás.

In [9]:
for c in ["Playing Time_MP", "Playing Time_Starts", "Playing Time_90s", "Starts_Compl"]:
    print(f"corr({c}, Playing Time_Min) = {df[c].corr(df['Playing Time_Min']):.3f}")

cols_volumen_redundante = ["Playing Time_MP", "Playing Time_Starts", "Playing Time_90s", "Starts_Compl"]
df = df.drop(columns=cols_volumen_redundante)
print(f"\nShape actual: {df.shape}")

corr(Playing Time_MP, Playing Time_Min) = 0.915
corr(Playing Time_Starts, Playing Time_Min) = 0.994
corr(Playing Time_90s, Playing Time_Min) = 1.000
corr(Starts_Compl, Playing Time_Min) = 0.879

Shape actual: (13023, 68)


## 5. Relativizar por posición las variables ofensivas y defensivas

5 goles es una temporada excepcional para un central pero mediocre para un delantero -- lo mismo al revés con entradas/intercepciones. Por eso re-escalamos `Gls`, `Ast`, `Sh`, `TklW` e `Int` con un `RobustScaler` calculado **dentro de cada posición**, no contra todo el dataset. Esto hay que hacerlo antes de que la sección 8 convierta `posicion_tm` en dummies.

(Para porteros estas estadísticas son casi siempre 0 -- `RobustScaler` lo maneja bien solo, no hace falta nada especial.)

In [10]:
from sklearn.preprocessing import RobustScaler

STATS_POR_POSICION = ["Performance_Gls", "Performance_Ast", "Standard_Sh", "Performance_TklW", "Performance_Int"]

partes = []
for pos, sub in df.groupby("posicion_tm", observed=True):
    rs = RobustScaler()
    escalado = rs.fit_transform(sub[STATS_POR_POSICION])
    partes.append(pd.DataFrame(escalado, columns=[f"{c}_rs" for c in STATS_POR_POSICION], index=sub.index))

df_pos_rs = pd.concat(partes).sort_index()
df[df_pos_rs.columns] = df_pos_rs  # sobreescribe Gls/Ast/Sh (antes globales) y crea TklW/Int (nuevas)

# Cada posición queda centrada en su propia mediana (~0), a diferencia del escalado global
print("Mediana de Performance_Gls_rs por posición (debería rondar 0 en todas):")
print(df.groupby("posicion_tm", observed=True)["Performance_Gls_rs"].median().round(2))

Mediana de Performance_Gls_rs por posición (debería rondar 0 en todas):
posicion_tm
Attacking Midfield    0.0
Central Midfield      0.0
Centre-Back           0.0
Centre-Forward        0.0
Defensive Midfield    0.0
Goalkeeper            0.0
Left Midfield         0.0
Left Winger           0.0
Left-Back             0.0
Right Midfield        0.0
Right Winger          0.0
Right-Back            0.0
Second Striker        0.0
Name: Performance_Gls_rs, dtype: float64


In [11]:
# Performance_Gls/Ast/Standard_Sh (crudos) los elimina la sección 6 más abajo (ya
# reemplazados por su _rs, ahora por posición). TklW/Int no tenían tratamiento previo:
df = df.drop(columns=["Performance_TklW", "Performance_Int"])
print(f"Shape actual: {df.shape}")

Shape actual: (13023, 68)


## 6. Una sola versión final por variable

El EDA ya dejó, para las variables más sesgadas, una versión `_rs` (robust-scaled) y/o `log1p_`. Mantener la original + las dos transformadas sería usar la misma señal tres veces. Elegimos una representación final por variable:

| Variable | Nos quedamos con | Se van |
|---|---|---|
| `Age`, `altura_m`, `Playing Time_Min`, `Performance_CrdY`, `Team Success_PPM` | su versión `_rs` | la original |
| `Performance_Gls`, `Ast`, `Standard_Sh` | su versión `_rs` | la original y el `log1p_` |
| `Standard_SoT`, `Performance_Crs`, `Performance_Fls` | su `log1p_` | la original |

Regla simple: si hay `_rs` disponible, usamos esa (preserva mejor la forma de la distribución); si solo hay `log1p_`, usamos esa.

In [12]:
cols_reemplazadas_por_rs = [
    "Age", "altura_m", "Playing Time_Min", "Performance_CrdY", "Team Success_PPM",
]
cols_reemplazadas_por_rs_y_log1p = ["Performance_Gls", "Performance_Ast", "Standard_Sh"]
cols_log1p_duplicado = [f"log1p_{c}" for c in cols_reemplazadas_por_rs_y_log1p]
cols_reemplazadas_por_log1p = ["Standard_SoT", "Performance_Crs", "Performance_Fls"]

df = df.drop(columns=(
    cols_reemplazadas_por_rs
    + cols_reemplazadas_por_rs_y_log1p
    + cols_log1p_duplicado
    + cols_reemplazadas_por_log1p
))
print(f"Shape actual: {df.shape}")

Shape actual: (13023, 54)


## 7. `Nation` se va

134 categorías, la mayoría con pocas filas. Convertirla en dummies metería más de 130 columnas dispersas y riesgo de sobreajuste, sin que tengamos una razón futbolística clara para que la nacionalidad (aparte de liga/equipo/rendimiento) explique el valor de mercado. La señal de "prestigio" que sí importa ya está en `Comp`, `Squad` y los indicadores de campeón.

In [13]:
print(f"Nation: {df['Nation'].nunique()} categorías")
df = df.drop(columns=["Nation"])
print(f"Shape actual: {df.shape}")

Nation: 134 categorías
Shape actual: (13023, 53)


## 8. Categóricas: posición, liga y equipo

`posicion_tm` (13 categorías) y `Comp` (5 ligas) son de baja cardinalidad, así que van directo a one-hot. `Squad` (130 equipos) necesita otro tratamiento -- ver 8.3.

**Posición:** usamos `posicion_tm` (más granular) y tiramos `Pos` de FBref, que mezcla combinaciones para jugadores multiposición.

**Liga:** el EDA ya mostró que Premier League y La Liga tienen medianas de valor más altas -- se codifica completa.

**Equipo, por qué no one-hot:** 130 categorías, muchas con 1-2 observaciones. Agrupar en "top-N + otros" (lo que hacíamos antes) pierde la señal de equipos medianos como Betis o Villarreal, que sí influyen en el valor de sus jugadores. Evaluamos tres alternativas:

| Método | Problema |
|---|---|
| Label Encoding | Inventa un orden entre equipos que no existe |
| Frequency Encoding | Mide tamaño de plantilla, no prestigio |
| **Target Encoding** | Usa el target -- hay que calcularlo con cuidado para no filtrar información |

Target Encoding es el que mejor sirve, pero **no se calcula acá** para no meter fuga de información: `Squad` queda sin codificar en este notebook, y el encoding se ajusta en `04-modelado.ipynb` solo sobre el train, después del split por jugador.

In [14]:
# 7.1 Posición: dummies desde posicion_tm, se elimina Pos (redundante, menos granular)
df = df.drop(columns=["Pos"])
dummies_pos = pd.get_dummies(df["posicion_tm"], prefix="pos", drop_first=True)

# 7.2 Competición: dummies desde Comp
dummies_comp = pd.get_dummies(df["Comp"], prefix="liga", drop_first=True)

print(f"Dummies de posición: {dummies_pos.shape[1]}")
print(f"Dummies de competición: {dummies_comp.shape[1]}")

df = pd.concat([df, dummies_pos, dummies_comp], axis=1)
df = df.drop(columns=["posicion_tm", "Comp"])

# 7.3 Equipo: se deja `Squad` como categórica sin codificar (130 categorías).
# El target encoding se ajusta en 04-modelado.ipynb solo con el train, tras el split.
df["Squad"] = df["Squad"].astype(str)
print(f"\nSquad se conserva sin codificar: {df['Squad'].nunique()} categorías (pendiente de target encoding en 04-modelado.ipynb)")

print(f"\nShape actual: {df.shape}")

Dummies de posición: 12
Dummies de competición: 4

Squad se conserva sin codificar: 130 categorías (pendiente de target encoding en 04-modelado.ipynb)

Shape actual: (13023, 66)


## 9. Vencimiento de contrato: versión final

`contrato_vence` es una fecha, no sirve tal cual para los modelos. La señal ya vive en `dias_contrato_restante` (del EDA), así que tiramos la fecha cruda. Falta decidir qué hacer con su 43% de nulos, porque ni la regresión lineal ni Random Forest toleran `NaN`. Imputamos con la mediana global (no asumimos nada sobre esos jugadores) y dejamos `flag_contrato_conocido` para que el modelo pueda aprender si la ausencia del dato ya dice algo por sí sola.

In [15]:
mediana_dias = df["dias_contrato_restante"].median()
df["dias_contrato_restante"] = df["dias_contrato_restante"].fillna(mediana_dias)
df = df.drop(columns=["contrato_vence"])

print(f"Mediana usada para imputar: {mediana_dias:.0f} días")
print(f"Nulos restantes en dias_contrato_restante: {df['dias_contrato_restante'].isna().sum()}")
print(f"Shape actual: {df.shape}")

Mediana usada para imputar: 1063 días
Nulos restantes en dias_contrato_restante: 0
Shape actual: (13023, 65)


## 10. Tipos finales y guardado

Convertimos las columnas booleanas a `0`/`1` y guardamos. `Squad` queda sin codificar a propósito (ver sección 8).

| Rol | Columnas |
|---|---|
| Identificadores | `Player`, `player_id`, `saison_id` |
| Objetivo | `valor_eur`, `log_valor_eur` (predecimos esta última) |
| Pendiente de codificar | `Squad` |
| Ya tratadas (`_rs`/`log1p_`) | edad, altura, minutos, goles, asistencias, tiros, amarillas, éxito de equipo, tiros a puerta, centros, faltas |
| Rendimiento sin redundancia | el resto de estadísticas crudas (penales, rojas, faltas recibidas, fuera de lugar, etc.) |
| Flags | sin tiros, sin titularidades, sin suplencias |
| Nuevas | `Ganador_Champions`, `Ganador_Liga` |
| Dummies | `pos_*`, `liga_*` |

Las variables de rendimiento que quedan sin tratar siguen correlacionando entre sí (el mismo cluster ofensivo/defensivo que vimos en el EDA), pero ninguna es duplicado exacto -- se recomienda revisar esto con regularización en el modelado.

In [16]:
bool_cols = df.select_dtypes(include="bool").columns
df[bool_cols] = df[bool_cols].astype(int)

print(f"Shape final: {df.shape}")
print(f"Columnas booleanas convertidas a int: {len(bool_cols)}")
print()
print("Nulos restantes:", df.isna().sum().sum())
print()
df.dtypes.value_counts()

Shape final: (13023, 65)
Columnas booleanas convertidas a int: 20

Nulos restantes: 0



int64      35
float64    27
str         3
Name: count, dtype: int64

In [17]:
from pathlib import Path

OUT_PATH = Path("../data/processed/fbref_tm_features.parquet")
df.to_parquet(OUT_PATH, index=False)
print(f"Dataset de variables para modelado guardado en: {OUT_PATH.resolve()}")
print(f"Shape: {df.shape}")

Dataset de variables para modelado guardado en: /home/diego/diplomado/data/processed/fbref_tm_features.parquet
Shape: (13023, 65)
